packages/libs/mcp-sdk/src/definitions.ts

```diff
@@ -0,0 +1,319 @@
+ *
+ * @param params - Array of Vincent parameter definitions
+ * @param addDelegatorPkpAddress - Whether to add the delegator eth address as a param
+ * @returns A Zod schema shape that can be used to create a validation schema
+ *
+ * @example
+ * ```typescript
+ * const parameters: VincentParameter[] = [
+ *   {
+ *     name: 'address',
+ *     type: 'address',
+ *     description: 'Ethereum address'
+ *   },
+ *   {
+ *     name: 'amount',
+ *     type: 'number',
+ *     description: 'Amount to transfer'
+ *   }
+ * ];
+ *
+ * const paramSchema = buildMcpParamDefinitions(parameters);
+ * const validationSchema = z.object(paramSchema);
+ * ```
+ */
+export function buildMcpParamDefinitions(
+  params: VincentParameter[],
+  addDelegatorPkpAddress: boolean
+): ZodRawShape {
+  const zodSchema = {} as ZodRawShape;
+
+  // Add the delegator PKP Eth address as a param. Delegatee is using the MCP and must specify which delegator to execute tools for
+  if (addDelegatorPkpAddress) {
+    zodSchema['pkpEthAddress'] = z
+      .string()
+      .describe(
+        "The delegator's PKP address that will execute the swap. For example 0xd8dA6BF26964aF9D7eEd9e03E53415D37aA96045."
+      );
+  }
+
+  // Add the rest of the parameters
+  params.forEach((param) => {
+    let paramZodSchema = ZodSchemaMap[param.type] || z.string();
+    if (param.optional) {
+      paramZodSchema = paramZodSchema.optional();
+    }
+    zodSchema[param.name] = paramZodSchema.describe(param.description);
+  });
+
+  return zodSchema;
+}
+
+export function buildMcpToolName(vincentAppDef: VincentAppDef, toolName: string) {
+  return `${vincentAppDef.name.toLowerCase().replace(' ', '-')}-V${vincentAppDef.version}-${toolName}`;
+}
+
+/**
+ * Creates a IPFS CID based Vincent Tool callback function to use in other contexts such as Agent Kits
+ *
+ * @param litNodeClient - The Lit Node client used to execute the tool
+ * @param delegateeSigner - The delegatee signer used to trigger the tool
+ * @param delegatorPkpEthAddress - The delegator to execute the tool in behalf of
+ * @param vincentToolDefWithIPFS - The tool definition with its IPFS CID
+ * @returns A callback function that executes the tool with the provided arguments
+ * @internal
+ */
+export function buildVincentToolCallback(
+  litNodeClient: LitNodeClient,
+  delegateeSigner: Signer,
+  delegatorPkpEthAddress: string | undefined,
+  vincentToolDefWithIPFS: VincentToolDefWithIPFS
+) {
+  return async (
+    args: ZodRawShape,
+    _extra?: RequestHandlerExtra<ServerRequest, ServerNotification>
+  ): Promise<{ success: boolean; error?: string; result?: object }> => {
+    try {
+      const sessionSigs = await generateVincentToolSessionSigs({
+        ethersSigner: delegateeSigner,
+        litNodeClient,
+      });
+      const executeJsResponse = await litNodeClient.executeJs({
+        ipfsId: vincentToolDefWithIPFS.ipfsCid,
+        sessionSigs,
+        jsParams: {
+          toolParams: args,
```

Comment:
This will pass in `pkpEthAddress`, which will be an unknown (invalid) property of `toolParams` for tools.

It should only be provided as a property of the [context jsParam](https://github.com/LIT-Protocol/Vincent/blob/d8f74db5eb0e23aa7d2cbefcae6cdc2f797847d3/packages/apps/tool-erc20-approval/src/lib/lit-action.ts#L11) which is a _sibling_ of `toolParams` -- but if I'm reading this right, this adds it to the `toolParams` directly.

Reply:
Yeah. If it was working on delegatee mode, then it would have been added to the mcp tool params and then passed to the vincent tool params

Removed it from params. The context remains the same

Reply:
This is property still being added in the current `main` version of this file, in two different places